src/annotator/integrations/pdf2htmlEx.tsx

```diff
@@ -0,0 +1,551 @@
+  }
+
+  sideBySideActive() {
+    return this._sideBySideActive;
+  }
+
+  /**
+   * Scroll to the location of an anchor in the PDF.
+   *
+   * If the anchor refers to a location that is an un-rendered page far from
+   * the viewport, then scrolling happens in three phases. First the document
+   * scrolls to the approximate location indicated by the placeholder anchor,
+   * then `scrollToAnchor` waits until the page's text layer is rendered and
+   * the annotation is re-anchored in the fully rendered page. Then it scrolls
+   * again to the final location.
+   */
+  async scrollToAnchor(anchor: Anchor) {
+    const annotation = anchor.annotation;
+    const inPlaceholder = anchorIsInPlaceholder(anchor);
+    const offset = this._anchorOffset(anchor);
+    if (offset === null) {
+      return;
+    }
+
+    // nb. We only compute the scroll offset once at the start of scrolling.
+    // This is important as the highlight may be removed from the document during
+    // the scroll due to a page transitioning from rendered <-> un-rendered.
+    await scrollElement(this.contentContainer(), offset);
+
+    if (inPlaceholder) {
+      const anchor = await this._waitForAnnotationToBeAnchored(
+        annotation,
+        this._reanchoringMaxWait,
+      );
+      if (!anchor) {
+        return;
+      }
+      const offset = this._anchorOffset(anchor);
+      if (offset === null) {
+        return;
+      }
+      await scrollElement(this.contentContainer(), offset);
+    }
+  }
+
+  /**
+   * Wait for an annotation to be anchored in a rendered page.
+   */
+  async _waitForAnnotationToBeAnchored(
+    annotation: AnnotationData,
+    maxWait: number,
+  ): Promise<Anchor | null> {
+    const start = Date.now();
+    let anchor;
+    do {
+      // nb. Re-anchoring might result in a different anchor object for the
+      // same annotation.
+      anchor = this._annotator.anchors.find(a => a.annotation === annotation);
+      if (!anchor || anchorIsInPlaceholder(anchor)) {
+        anchor = null;
+
+        // If no anchor was found, wait a bit longer and check again to see if
+        // re-anchoring completed.
+        await delay(20);
+      }
+    } while (!anchor && Date.now() - start < maxWait);
+    return anchor ?? null;
+  }
+
+  /**
+   * Return the offset that the PDF content container would need to be scrolled
+   * to, in order to make an anchor visible.
+   *
+   * @return - Target offset or `null` if this anchor was not resolved
+   */
+  _anchorOffset(anchor: Anchor): number | null {
+    if (!anchor.highlights) {
+      // This anchor was not resolved to a location in the document.
+      return null;
+    }
+    const highlight = anchor.highlights[0];
+
+    // PdfEX hides page internals using display: none, so scroll to the page instead
+    if (! highlight.checkVisibility({ checkVisibilityCSS: true })) {
+      const page = highlight.closest('.pf') as HTMLElement;
+      return offsetRelativeTo(page, this.contentContainer());
```

Comment:
This is the best solution I could find for navigating to highlights that are hidden by pdf2HtmlEx. Adding a placeholder here is tricky because we have to remove the highlights before we can later replace the placeholder with highlights again.

Reply:
Perhaps we can just unhide the element as well